specs/Slots.md

```diff
@@ -0,0 +1,156 @@
+    {/* renders a <Loader /> element */}
+    <slots.loader />
+    {/* renders a <span /> element */}
+    <slots.icon />
+  </slots.root>
+</>
+```
+
+A default set of components is manually configured per each component, but is optional, by default slots are represented by `div` elements.
+
+## Slots usage
+
+There are several forms of values that can be provided, but all of them share one common thing - they will produce a proper layout inside a component.
+
+### An object as a value
+
+If you will pass an object to component's props we will handle it as props for a slot, this behavior can be considered as declaring a JSX element with a javascript object. In an example below, props will be passed to an `icon` slot:
+
+```jsx
+<Button icon={{ children: <FooIcon />, className: 'an-awesome-slot', id: '#button-icon' }} />
+```
+
+```html
+<!-- 💡 Simplified DOM output -->
+<button class="ms-Button">
+  <!-- 👇 An additional class and id have been added to markup -->
+  <span class="ms-Button-icon an-awesome-slot" id="#button-icon">
+    <!-- 👇 An icon will be rendered inside a slot as it is passed as children -->
+    <span class="ms-Icon"><svg /></span>
+  </span>
+</button>
+```
+
+Any props what are valid for a component can be passed to a slot, i.e. `data-*` attributes and event handlers:
+
+```jsx
+<Button
+  icon={{
+    children: <FooIcon />,
+    'data-test-id': 'button-foo-icon',
+    onClick: () => {
+      console.log('A click on an icon slot!');
+    },
+  }}
+/>
+```
+
+### Primitives as a value
+
+We consider JSX elements as primitive values that will be passed to a meaningful prop, by default to `children` (but this can be configured in the component's implementation):
+
+```jsx
+<>
+  <Button icon={{ children: <FooIcon /> }} />
+  {/* 💡 will produce the same JSX/HTML markup, as will be expanded to { children: <FooIcon /> } */}
+  <Button icon={<FooIcon />} />
+</>
+```
+
+Such usage of slots is called shorthands, similarly to [CSS shorthand properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties). The same behavior is applicable strings and numbers:
+
+```jsx
+<>
+  <Button icon={{ children: 'text instead of icon' }} />
+  {/* 💡 will produce the same JSX/HTML markup */}
+  <Button icon="text instead of icon" />
+</>
+```
+
+To disable slot rendering you can use falsy (`null`, `false`) values:
+
+```jsx
+// 💡 A loader slot will be hidden
+<>
+  <Button loading loader={null} />
+  <Button loading loader={false} />
+</>
+```
+
+`true` will render a slot without any content:
+
+```jsx
+// 💡 A loader slot will be renderer without any content
+<Button loading loader={true} />
+```
+
+```html
+<!-- 💡 Simplified DOM output -->
+<button class="ms-Button">
+  <span class="ms-Button-loader"></span>
+</button>
+```
+
+### Renders props via `children` function
+
+Slots can be deeply customized via `children` function that behave similarly to [Render Props pattern](https://reactjs.org/docs/render-props.html) in React:
```

Comment:
I think that this is the actual goal of the `children` render func right ? to not have to throw away the original result

```suggestion
Slots can be deeply customized without ignoring the original content by using `children` as a function that behave similarly to [Render Props pattern](https://reactjs.org/docs/render-props.html) in React:
```